### PR TITLE
Partially revert "fix voting on multiple roles behavior description" due to CVE-2020-5275

### PR DESCRIPTION
**Pull request by ajgarlag**
This partially reverts commit 4b76554f55ec3166f1b0dc8ee9ba628f80842051.

See https://github.com/symfony/symfony/commit/c935e4a3fba6cc2ab463a6ca382858068d63cebf

### CODE REVIEW

File: security/access_control.rst [hunk 1]
```diff
@@ -157,7 +157,9 @@ options:
 * ``roles`` If the user does not have the given role, then access is denied
   (internally, an :class:`Symfony\\Component\\Security\\Core\\Exception\\AccessDeniedException`
   is thrown). If this value is an array of multiple roles, the user must have
-  at least one of them.
+  at least one of them (when using the default ``affirmative`` strategy in the
+  :ref:`Access Decision Manager <components-security-access-decision-manager>`)
+  or all of them when using the ``unanimous`` strategy;
```
Comment (dbrumann) [inline security/access_control.rst]: I like the idea, but can we change the formatting to something like:
```suggestion
  * at least one of them when using the default ``affirmative`` strategy
  * all of them when using the ``unanimous`` strategy
in the :ref:`Access Decision Manager <components-security-access-decision-manager>`.
```

**Review by dbrumann — APPROVED** [review]

**Review by dbrumann — APPROVED** [review]
:+1: I like it. Thanks.

File: security/access_control.rst [hunk 2]
```diff
@@ -156,8 +156,13 @@ options:
 
 * ``roles`` If the user does not have the given role, then access is denied
   (internally, an :class:`Symfony\\Component\\Security\\Core\\Exception\\AccessDeniedException`
-  is thrown). If this value is an array of multiple roles, the user must have
-  at least one of them.
+  is thrown). If this value is an array of multiple roles, the user must have:
+  
+  * at least one of them when using the default ``affirmative`` strategy.
+  * more granted than denied roles when using the ``consensus`` strategy.
```
Comment (dbrumann) [inline security/access_control.rst]: Depending on the config it could also be "at least as many granted as denied", but that's probably a bit pedantic :D

Comment (ajgarlag) [reply security/access_control.rst]: I know, so revised `affirmative` description in https://github.com/symfony/symfony-docs/blob/4.4/components/security/authorization.rst#access-decision-manager

**Comment by javiereguiluz** [comment]
This is security-related, so we'd like to merge this as soon as possible. @wouterj could you please check these proposed changes? Thanks!

**Comment by javiereguiluz** [comment]
After #15503 is merged, we'll need to check if this is still relevant and if we need to fix it. Thanks!

**Comment by ajgarlag** [comment]
> After #15503 is merged, we'll need to check if this is still relevant and if we need to fix it. Thanks!

@javiereguiluz #15503 is merged and I think this is still relevant. I've rebased the PR.

**Comment by wouterj** [comment]
Hi @ajgarlag!

I'm sorry for the long wait here. I honestly wasn't sure what to do with these changes, as I do see your point about "roles" (and also being able to pass an array of attributes to this setting).

For now, I've decided to follow @xabbuh's suggestion. The reason is that an array of roles does not have a strict OR or AND relation, nor does this depend on the strategy. For full information:

* The strategy determines how the votes of all voters are merged. E.g. `RoleVoter` might return DENIED, whereas `AuthenticatedVoter` returns GRANTED.
* However, the list or roles is provided to each voter and the voter itself determines how to turn the list of roles into a single vote. So if you have multiple attributes that are handled by the same voter (e.g. 2 `ROLE_*` attributes), the strategy is no longer important and the relationship is hardcoded in the voter (in the case of roles without using role_hierarchy, it's OR).
* For roles, this becomes even more confusing as plain roles are handled by `RoleVoter`, but hierarchical roles are handled by `RoleHierarchyVoter`. So if a ROLE_ADMIN is also a ROLE_USER (through the hierarchy) and you use `['ROLE_USER', 'ROLE_ADMIN']`, you end up with a completely confusing situation (where both voters merge the attributes into a single vote using hardcoded logic, and the Access decision manager merges the two votes based on the strategy).

This is exactly the reason why we removed voting on more than one attribute at the same time. Unfortunately, it's the worst change I've ever contributed to Symfony (implementation wise), which means we still have the "roles" setting and ability to use an array of roles in `access_control`. Let's see if we can fix this for 6.x :)

Anyway, in case you're still reading this message: Thanks for remaining active in this PR! If you have any questions, feel free to comment :)

**Pull request merged**